File: src/main/java/net/kigawa/kutil/unit/UnitContainer.java
package net.kigawa.kutil.unit;

import kotlin.Metadata;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.net.JarURLConnection;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

public class UnitContainer
{
    private final UnitContainer parent;
    private final UnitsMap unitInfoMap = new UnitsMap();

    public UnitContainer(Object... units) {
        this(null, units);
    }

    public UnitContainer(UnitContainer parent, Object... units) {
        this.parent = parent;
        registerUnit(this);

        for (Object unit : units) {
            registerUnit(unit);
        }
    }

    public void loadUnits(Class<?> rootClass) throws UnitException {
        registerUnits(rootClass);
        initUnits();
    }

    public <T> void registerUnit(T unit) {
        var containerInfo = new UnitInfo(unit.getClass());
        containerInfo.unit = unit;
        unitInfoMap.put(unit.getClass(), containerInfo);
    }

    public <T> T getUnit(Class<T> unitClass) {
        var unitInfo = unitInfoMap.get(unitClass);
        if (unitInfo != null) {
            if (unitInfo.unit != null) return (T) unitInfo.unit;
            try {
                return initUnit(unitClass);
            } catch (UnitException e) {
                throw new RuntimeUnitException("could not init unit: " + unitClass, e);
            }
        }

        if (parent != null) return parent.getUnit(unitClass);

        throw new RuntimeUnitException("unit is not found: " + unitClass);
    }

    public Set<Class<?>> getAllClasses() {
        return unitInfoMap.keySet();
    }

    private void initUnits() throws UnitException {
        var exceptions = new LinkedList<Exception>();
        for (var unitClass : unitInfoMap.keySet()) {
            try {
                initUnit(unitClass);
            } catch (Exception e) {
                exceptions.add(new UnitException("could not init unit: " + unitClass, e));
            }
        }
        throwExceptions(exceptions, new UnitException("there are exceptions when init units"));
    }

    private <T> T initUnit(Class<T> unitClass) throws UnitException {
        var unitInfo = unitInfoMap.get(unitClass);

        if (unitInfo == null) throw new UnitException("could not find unit: " + unitClass);
        if (unitInfo.unit != null) return (T) unitInfo.unit;

        if (unitClass.isAnnotationPresent(Metadata.class)) {
            unitInfo.unit = initKotlinClass(unitClass, unitInfo);
            return (T) unitInfo.unit;
        }
        unitInfo.unit = initNormalClass(unitClass, unitInfo);
        return (T) unitInfo.unit;
    }

    private Object initKotlinClass(Class<?> unitClass, UnitInfo unitInfo) throws UnitException {
        try {
            var field = unitClass.getField("INSTANCE");

            return field.get(null);
        } catch (NoSuchFieldException e) {
            return initNormalClass(unitClass, unitInfo);
        } catch (IllegalAccessException e) {
            throw new UnitException("could not access INSTANCE field: ", e);
        }
    }

    private Object initNormalClass(Class<?> unitClass, UnitInfo unitInfo) throws UnitException {
        var constructor = unitInfo.getConstructor();

        var parameters = constructor.getParameterTypes();
        var objects = new Object[parameters.length];
        for (int i = 0; i < parameters.length; i++) {
            objects[i] = getUnit(parameters[i]);
        }
        try {
            return constructor.newInstance(objects);
        } catch (InstantiationException | IllegalAccessException e) {
            throw new UnitException("could not init unit: " + unitClass, e);
        } catch (InvocationTargetException e) {
            throw new UnitException("could not init unit: " + unitClass, e.getCause());
        }
    }

    private void registerUnits(Class<?> rootClass) throws UnitException {
        unitInfoMap.clearCache();
        var rootPackage = rootClass.getPackage();
        var resourceName = rootPackage.getName().replace('.', '/');
        var classLoader = rootClass.getClassLoader();
        var root = classLoader.getResource(resourceName);
        var exceptions = new LinkedList<Exception>();

        if (root == null) throw new UnitException("could not load class files");

        if ("file".equals(root.getProtocol())) {
            var files = new File(root.getFile()).listFiles((dir, name) -> name.endsWith(".class"));
            if (files == null) throw new UnitException("cold not load unit files");

            for (var file : files) {
                var name = file.getName();
                name = name.replaceAll(".class$", "");
                name = rootPackage.getName() + "." + name;

                try {
                    registerUnit(Class.forName(name));
                } catch (Exception e) {
                    exceptions.add(new UnitException("cold not load unit: " + name, e));
                }
            }
        } else if ("jar".equals(root.getProtocol())) {
            try (var jarFile = ((JarURLConnection) root.openConnection()).getJarFile()) {
                for (var entry : Collections.list(jarFile.entries())) {
                    var name = entry.getName();
                    if (!name.startsWith(resourceName)) continue;
                    if (!name.endsWith(".class")) continue;
                    name = name.replace('/', '.').replaceAll(".class$", "");

                    try {
                        registerUnit(Class.forName(name));
                    } catch (Exception e) {
                        exceptions.add(new UnitException("could not load unit: " + name, e));
                    }
                }
            } catch (IOException e) {
                throw new UnitException("could not load units file", e);
            }
        }
        throwExceptions(exceptions, new UnitException("there are exceptions when load units"));
    }

    private void registerUnit(Class<?> unitClass) {
        if (unitClass.getAnnotation(Unit.class) == null) return;
        unitInfoMap.put(unitClass, new UnitInfo(unitClass));
    }

    private <E extends Exception> void throwExceptions(List<Exception> exceptions, E base) throws E {
        if (exceptions.size() == 0) return;

        for (var e : exceptions) {
            base.addSuppressed(e);
        }
        throw base;
    }
}

File: src/main/java/net/kigawa/kutil/unit/UnitContainer.kt
package net.kigawa.kutil.unit

import net.kigawa.kutil.unit.runtimeexception.RuntimeUnitException
import net.kigawa.kutil.unit.runtimeexception.UnitNotInitException
import java.io.File
import java.io.IOException
import java.lang.reflect.InvocationTargetException
import java.net.JarURLConnection
import java.net.URL
import java.util.*

class UnitContainer(
    private val parent: UnitContainer? = null,
    vararg units: Any,
)
{
    constructor(vararg units: Any) : this(null, *units)

    private val unitInfoMap = UnitsMap()

    init
    {
        registerUnit(this)
        for (unit in units)
        {
            registerUnit(unit)
        }
    }

    @Throws(UnitException::class)
    fun loadUnits(rootClass: Class<*>)
    {
        registerUnits(rootClass)
        initUnits()
    }

    fun <T : Any> registerUnit(unit: T)
    {
        val containerInfo = UnitInfo(unit.javaClass)
        containerInfo.unit = unit
        unitInfoMap.put(unit.javaClass, containerInfo)
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> getUnit(unitClass: Class<T>): T?
    {
        val unitInfo = unitInfoMap.get(unitClass)
        if (unitInfo != null)
        {
            if (unitInfo.unit != null) return unitInfo.unit as T
            throw UnitNotInitException("unit is not initialized")
        }
        if (parent != null) return parent.getUnit(unitClass)
        throw RuntimeUnitException("unit is not found: $unitClass")
    }

    private fun <T> getUnitAndInit(unitClass: Class<T>): T?
    {
        val unitInfo = unitInfoMap.get(unitClass)
        if (unitInfo != null)
        {
            return try
            {
                getUnit(unitClass)
            } catch (e: UnitNotInitException)
            {
                initUnit(unitClass)
            }
        }
        if (parent != null) return parent.getUnitAndInit(unitClass)
        throw RuntimeUnitException("unit is not found: $unitClass")
    }

    val allClasses: Set<Class<*>>
        get() = unitInfoMap.keySet()

    fun initUnits()
    {
        val exceptions = LinkedList<Exception>()
        for (unitClass in unitInfoMap.keySet())
        {
            try
            {
                initUnit(unitClass)
            } catch (e: Throwable)
            {
                exceptions.add(UnitException("could not init unit: $unitClass", e))
            }
        }
        throwExceptions(exceptions, RuntimeUnitException("there are exceptions when init units"))
    }

    @Suppress("UNCHECKED_CAST")
    @Throws(UnitException::class)
    private fun <T> initUnit(unitClass: Class<T>): T
    {
        val unitInfo = unitInfoMap.get(unitClass) ?: throw UnitException("could not find unit: $unitClass")
        if (unitInfo.unit != null) return unitInfo.unit as T
        if (unitClass.isAnnotationPresent(Metadata::class.java))
        {
            unitInfo.unit = initKotlinClass(unitClass, unitInfo)
            return unitInfo.unit as T
        }
        unitInfo.unit = initNormalClass(unitClass, unitInfo)
        return unitInfo.unit as T
    }

    @Throws(UnitException::class)
    private fun initKotlinClass(unitClass: Class<*>, unitInfo: UnitInfo): Any
    {
        return try
        {
            val field = unitClass.getField("INSTANCE")
            field[null]
        } catch (e: NoSuchFieldException)
        {
            initNormalClass(unitClass, unitInfo)
        } catch (e: IllegalAccessException)
        {
            throw UnitException("could not access INSTANCE field: ", e)
        }
    }

    @Throws(UnitException::class)
    private fun initNormalClass(unitClass: Class<*>, unitInfo: UnitInfo): Any
    {
        val constructor = unitInfo.constructor
        val parameters = constructor.parameterTypes
        val objects = arrayOfNulls<Any>(parameters.size)
        for (i in parameters.indices)
        {
            objects[i] = getUnitAndInit(parameters[i])
        }
        return try
        {
            constructor.newInstance(*objects)
        } catch (e: InstantiationException)
        {
            throw UnitException("could not init unit: $unitClass", e)
        } catch (e: IllegalAccessException)
        {
            throw UnitException("could not init unit: $unitClass", e)
        } catch (e: InvocationTargetException)
        {
            throw UnitException("could not init unit: $unitClass", e.cause)
        }
    }

    fun registerUnits(rootClass: Class<*>)
    {
        unitInfoMap.clearCache()
        val rootPackage = rootClass.getPackage()
        val classLoader = rootClass.classLoader
        val root = classLoader.getResource(rootPackage.name.replace('.', '/'))
        val exceptions = LinkedList<Exception>()
        if (root == null) throw RuntimeUnitException("could not load class files")
        if ("file" == root.protocol)
        {
            registerUnit(File(root.file), rootPackage.name)
        } else if ("jar" == root.protocol)
        {
            registerUnit(root, rootPackage.name)
        }
        throwExceptions(exceptions, RuntimeUnitException("there are exceptions when load units"))
    }

    private fun registerUnit(root: URL, packageName: String): MutableList<Throwable>
    {
        val exceptions = mutableListOf<Throwable>()
        try
        {
            (root.openConnection() as JarURLConnection).jarFile.use { jarFile ->
                for (entry in Collections.list(jarFile.entries()))
                {
                    var name = entry.name
                    if (!name.startsWith(packageName.replace('.', '/'))) continue
                    if (!name.endsWith(".class")) continue
                    name = name.replace('/', '.').replace(".class$".toRegex(), "")
                    try
                    {
                        registerUnit(Class.forName(name))
                    } catch (e: Exception)
                    {
                        exceptions.add(UnitException("could not load unit: $name", e))
                    }
                }
            }
        } catch (e: IOException)
        {
            exceptions.add(RuntimeUnitException("could not load units file", e))
        }
        return exceptions
    }


    private fun registerUnit(dir: File, packageName: String): MutableList<Throwable>
    {
        val exceptions = mutableListOf<Throwable>()
        for (file in dir.listFiles() ?: throw UnitException("cold not load unit files"))
        {
            if (file.isDirectory)
            {
                registerUnit(file, packageName + "." + file.name)
                continue
            }
            if (!file.name.endsWith(".class")) continue
            var name = file.name
            name = name.replace(".class$".toRegex(), "")
            name = "$packageName.$name"
            try
            {
                registerUnit(Class.forName(name))
            } catch (e: Exception)
            {
                exceptions.add(UnitException("cold not load unit: $name", e))
            }
        }
        return exceptions
    }

    private fun registerUnit(unitClass: Class<*>)
    {
        val annotation = unitClass.getAnnotation(Unit::class.java) ?: return
        unitInfoMap.put(unitClass, UnitInfo(unitClass, annotation))
    }

    @Throws(Throwable::class)
    private fun <E : Throwable> throwExceptions(exceptions: List<Exception>, base: E)
    {
        if (exceptions.isEmpty()) return
        exceptions.forEach {
            base.addSuppressed(it)
        }
        throw base
    }
}